psycopg_c/psycopg_c/types/datetime.pyx

```diff
@@ -971,11 +971,11 @@ cdef class IntervalBinaryLoader(CLoader):
     format = PQ_BINARY
 
     cdef object cload(self, const char *data, size_t length):
-        cdef int32_t bedata[4]
+        cdef int64_t bedata[2]
         memcpy(&bedata, data, sizeof(bedata))
-        cdef int64_t val = endian.be64toh((<uint64_t *>bedata)[0])
-        cdef int32_t days = endian.be32toh(bedata[2])
-        cdef int32_t months = endian.be32toh(bedata[3])
+        cdef int64_t val = endian.be64toh(bedata[0])
+        cdef int32_t days = endian.be32toh((<uint32_t *>bedata)[2])
```

Comment:
@dvarrazzo I think these lines are still UB, you cannot cast a `int64_t*` to a `uint32_t*` and dereference it.

Reply:
here's my suggestion:
declare `bedata` as an object of type
```
struct {
  uint64_t val;
  uint32_t days;
  uint32_t months;
};
```
and then initialize it with your call to `memcpy`

Reply:
@SoapGentoo I understand that the problem comes from unaligned memory. So I understand if aligning 64 bits on 32 bits boundaries can get unaligned, but aligning 32 bits on the 64 bits boundary doesn't. The compiler doesn't report any error. Is there a stricter check you suggest to add?

I'm not sure about the struct; if any, I would use two separate memcpy:

```diff
diff --git a/psycopg_c/psycopg_c/types/datetime.pyx b/psycopg_c/psycopg_c/types/datetime.pyx
index 44b227ed..e2a6ac8a 100644
--- a/psycopg_c/psycopg_c/types/datetime.pyx
+++ b/psycopg_c/psycopg_c/types/datetime.pyx
@@ -971,11 +971,13 @@ cdef class IntervalBinaryLoader(CLoader):
     format = PQ_BINARY
 
     cdef object cload(self, const char *data, size_t length):
-        cdef int64_t bedata[2]
-        memcpy(&bedata, data, sizeof(bedata))
-        cdef int64_t val = endian.be64toh(bedata[0])
-        cdef int32_t days = endian.be32toh((<uint32_t *>bedata)[2])
-        cdef int32_t months = endian.be32toh((<uint32_t *>bedata)[3])
+        cdef int64_t beval
+        cdef int32_t bedm[2]
+        memcpy(&beval, data, sizeof(beval))
+        memcpy(bedm, data + sizeof(beval), sizeof(bedm))
+        cdef int64_t val = endian.be64toh(beval)
+        cdef int32_t days = endian.be32toh(bedm[0])
+        cdef int32_t months = endian.be32toh(bedm[1])
 
         cdef int years
         with cython.cdivision(True):
```

(this actually makes me notice that the original line has an error: it should have been `memcpy(bedata, ...` not `memcpy(&bedata, ...`. How does it even work?)

Reply:
I would be more comfortable making this change than using a struct. I don't know the rules for padding in their most intimate details, but what guarantees that some compiler won't align the data to 64 bits and add 4 bytes of padding between days and months?

Reply:
@dvarrazzo the alignment argument is correct, but `-Wstrict-aliasing` is more than just alignment. You cannot in general cast a `T*` to a `U*` and then dereference it, unless a very specific and constrained set of circumstances is true (for instance, `U = void` or `U = {char, unsigned char}` or some other rules that don't matter here).

Your double-copy is just as fine, if not clearer, so I would go with that route.
`memcpy(bedata, ...` works, because in C arrays often decay into pointers, and since the first argument of `memcpy` is a `void*`, the compiler implicitly converts the array to a pointer to its first element (which is absolutely fine and well defined).

Reply:
Thank you for the feedback, @SoapGentoo. I've pushed a change in #879 for testing.